### ISSUE TITLE
test_target_defaultmap calls CPU functions in a target region

### ISSUE THREAD

**Issue by naromero77intel** [SOLLVE/sollve_vv#545]
**Describe the bug**
I believe this commit:
https://github.com/SOLLVE/sollve_vv/commit/fffc1fcb6d308efad7ef85a2d763717f41078f36

Introduced a bug into test_target_defaultmap.

The bug is in `test_defaultmap_off` part which tests firstprivate on scalars. `test_defaultmap_on` is fine.

With the Intel oneAPI Fortran compiler (ifx), the test seems to "sort of run." Until you actually try to isolate the problem further and then you get seg. faults and other weird errors. 

The issue is easy to overlook:
https://github.com/SOLLVE/sollve_vv/blob/master/tests/4.5/target/test_target_defaultmap.F90#L141-L142
https://github.com/SOLLVE/sollve_vv/blob/master/tests/4.5/target/test_target_defaultmap.F90#L146

`same_val_real32`, `same_val_real64`, and `same_val_cmplx` are functions that are defined on the CPU, but the tests calls them in a target region which is NOT allowed unless the function is mapped to the device with `omp declare target`. (Though I suppose this could possibly work in some future when compilers support reverse offloading).

I would recommend re-writing the test so that those functions are NOT called in the target region for `test_defaultmap_off` part of the test.

**Test that it applies to**
test_target_defaultmap

**To Reproduce**
```
ifx -o test_target_defaultmap.x -fiopenmp -fopenmp-targets=spir64 test_target_defaultmap.F90
```

I created a smaller reproducer from this test case, but I don't think it would be that useful to you.

**Expected behavior**
The test should pass once you get rid of those functions that are called in the target region.

**Compiler**
Intel oneAPI compiler (latest internal version)

**Accelerator hardware**
Intel Integrate GPU

**Comment by tob2** [comment]
> same_val_real32, same_val_real64, and same_val_cmplx are functions that are defined on the CPU, but the tests calls them in a target region which is NOT allowed unless the function is mapped to the device with omp declare target.
> (Though I suppose this could possibly work in some future when compilers support reverse offloading)

Let's start with the parenthesis part: I think that would be a bug – reverse offload should only happen with an explicit `!$omp target device(ancestor:1)` and not implicitly. I also do not think that reverse offloading is hard – for both reverse offload and I/O from target regions, essentially the same mechanism is used. And I/O from within target regions is widely supported.

Regarding the first, non-parenthesized part: I think you are right, but only because this is a 4.5 testcase. Since OpenMP 5.0, the following applies. _(That is also the reason it works with all tested compilers.)_
> "If a function (C, C++, Fortran) or subroutine (Fortran) is referenced in a target construct then that function or subroutine is treated as if its name had appeared in a to clause on a declare target directive." (OpenMP 5.0, "2.12.5 target Construct", [173:1–3]).

And in this case, the procedure is not only in the same file – but even in the same _compilation unit_ (→ OpenMP glossary). Thus, I think you have found a compiler bug – unless the compiler lacks most of OpenMP 5.0.

In terms of the testsuite, given that it a 4.5 testcase, I concur that a `declare target` should be added. Sorry that I missed it – I have read too much OpenMP 5.x lately and forgot this when fixing sollve_vv issues.

**Comment by spophale** [comment]
Thanks @naromero77intel and @tob2 ! Yes, I agree with Tobias. Will add a fix soon.

**Comment by tmh97** [comment]
Fixed with #546, closing this issue. Thanks @tob2 and @naromero77intel for your help!
